src/components/Products/Products.js

```diff
@@ -0,0 +1,302 @@
+import { Link } from "react-router-dom";
+import { useProduct } from "../../context/product-context";
+import "./products.css";
+import { useOffers } from "../../context/offers-context";
+import { BsLightningFill } from "react-icons/bs";
+import { GiRoundStar } from "react-icons/gi";
+import { initialData, reducerFunction } from "../../reducer/product-reducer";
+import {
+  getPriceSortedData,
+  getRatingSortedData,
+  getFilteredData,
+} from "../Utilities/product-utilities";
+import { AddToCart } from "../Cart/AddToCart";
+import { AddToWishlist } from "../Wishlist/AddToWishlist";
+
+export const Products = () => {
+  const { productList } = useProduct();
+  const { offersList } = useOffers();
+  const [showFilters, setShowFilters] = useState(false);
+  const [sliderVal, setSliderVal] = useState(0);
+
+  const [
+    { inStockOnly, fastDeliveryOnly, sortByPrice, sortByRating, sortByOffers },
+    dispatch,
+  ] = useReducer(reducerFunction, initialData);
+
+  const getOfferSortedData = (existingProductList, sortByOffers) => {
+    const noDiscount = offersList.find((offer) => offer.name === "No Discount");
+    const tenPercent = offersList.find(
+      (offer) => offer.name === "Ten Percent Discount"
+    );
+    const twelvePercent = offersList.find(
+      (offer) => offer.name === "Twelve Percent"
+    );
+
+    if (sortByOffers === noDiscount) {
+      return existingProductList.filter(
+        (product) => product.offers[0] === noDiscount._id
+      );
+    } else if (sortByOffers === tenPercent) {
+      return existingProductList.filter(
+        (product) => product.offers[0] === tenPercent._id
+      );
+    } else if (sortByOffers === twelvePercent) {
+      return existingProductList.filter(
+        (product) => product.offers[0] === twelvePercent._id
+      );
+    } else return existingProductList;
+  };
+
+  const priceSortedData = getPriceSortedData(productList, sortByPrice);
+  const ratingSortedData = getRatingSortedData(priceSortedData, sortByRating);
+  const offersSortedData = getOfferSortedData(
+    ratingSortedData,
+    offersList,
+    sortByOffers
+  );
+  const filteredData = getFilteredData(
+    offersSortedData,
+    fastDeliveryOnly,
+    inStockOnly
+  );
+
+  const filtersDiv = useRef("none");
+  const sliderStars = useRef(0);
+
+  const getSliderValue = () => {
+    setSliderVal(() => sliderStars.current.value);
+  };
+
+  const toggleFilterDiv = () => {
+    setShowFilters((filters) => !filters);
+    const showVal = showFilters ? "flex" : "none";
+    filtersDiv.current.style.display = showVal;
+  };
+
+  useEffect(() => dispatch({ type: "SORT_BY_RATING", payload: sliderVal }), [
+    sliderVal,
+    setSliderVal,
+  ]);
+
+  const addRatingStars = (rating) => {
+    let starString = [];
+    for (let i = 0; i < rating; i++) {
+      starString.push(
+        <GiRoundStar size={20} className="mg-r-025 txt-yellow" />
+      );
+    }
+    return starString;
+  };
+
+  return (
+    <>
```

Comment:
You can shift the filters display section to a different component. This will make the products page a little cleaner and shorter. It's just a suggestion.

Reply:
will do that! thanks!